vpr/src/route/route_budgets.cpp

```diff
@@ -150,50 +172,74 @@ void route_budgets::calculate_delay_targets(ClusterNetId net_id, ClusterPinId pi
     int ipin = cluster_ctx.clb_nlist.pin_net_index(pin_id);
 
     /*Target delay is calculated using equation in the RCV algorithm*/
-    delay_target[net_id][ipin] = std::min(0.5 * (delay_min_budget[net_id][ipin] + delay_max_budget[net_id][ipin]), delay_min_budget[net_id][ipin] + 0.1e-9);
+    if (delay_min_budget[net_id][ipin] != 0) {
+        delay_target[net_id][ipin] = std::min(0.5 * (delay_min_budget[net_id][ipin] + delay_max_budget[net_id][ipin]), delay_min_budget[net_id][ipin] + 0.1e-9);
+    } else {
+        delay_target[net_id][ipin] = 0;
+    }
 }
 
-void route_budgets::allocate_slack_using_weights(ClbNetPinsMatrix<float>& net_delay, const ClusteredPinAtomPinsLookup& netlist_pin_lookup) {
+void route_budgets::allocate_slack_using_weights(ClbNetPinsMatrix<float>& net_delay, const ClusteredPinAtomPinsLookup& netlist_pin_lookup, bool negative_hold_slack) {
     /*The minimax PERT algorithm uses a weight based approach to allocate slack for each connection
      * The formula used where c is a connection is
      * slack_allocated(c) = (slack(c)*weight(c)/max_weight_of_all_path_through(c)).
      * Weights here are defined as the delay for the connections
      * Values for conditions in the while loops are pulled from the RCV paper*/
     std::shared_ptr<SetupHoldTimingInfo> timing_info = nullptr;
+    std::shared_ptr<SetupHoldTimingInfo> timing_info_min = nullptr;
+    std::shared_ptr<SetupHoldTimingInfo> original_timing_info = nullptr;
 
     unsigned iteration;
     float max_budget_change;
 
     /*Preprocessing algorithm in order to consider short paths when setting initial maximum budgets.
      * Not necessary unless budgets are really hard to meet*/
-    //process_negative_slack_using_minimax();
+    // process_negative_slack_using_minimax();
+    if (negative_hold_slack) {
+        process_negative_slack_using_minimax(net_delay, netlist_pin_lookup);
+    }
 
     iteration = 0;
     max_budget_change = 900e-12;
 
+    original_timing_info = perform_sta(net_delay);
+
     /*This allocates long path slack and increases the budgets*/
-    while ((iteration > 3 && max_budget_change > 800e-12) || iteration <= 3) {
+    while ((iteration > 3 && max_budget_change > 5e-12) || iteration <= 3) {
```

Comment:
Still have a lot of magic here?

Reply:
You're right, I'll constantize route_budgets right now